### PR TITLE
[sentry-crashpad/xxx] conan v2 migration

### PR DESCRIPTION
**Pull request by MartinDelille**
My first attempt in a conan v2 migration.
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/sentry-crashpad/all/conanfile.py [hunk 1]
```diff
@@ -1,7 +1,11 @@
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
 import os
 
+from conan import ConanFile
+from conan.tools.scm import Version
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.files import get, apply_conandata_patches, copy, rmdir, rm, replace_in_file
+from conan.tools.apple.apple import is_apple_os
+
```
Comment (uilianries) [inline recipes/sentry-crashpad/all/conanfile.py:8]:
```suggestion
from conan.errors import ConanInvalidConfiguration

```

File: recipes/sentry-crashpad/all/conanfile.py [hunk 2]
```diff
@@ -1,7 +1,12 @@
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
 import os
 
+from conan import ConanFile
+from conan.tools.scm import Version
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
```
Comment (uilianries) [inline recipes/sentry-crashpad/all/conanfile.py:5]:
```suggestion
from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
```

File: recipes/sentry-crashpad/all/conanfile.py [hunk 3]
```diff
@@ -53,14 +48,24 @@ def _minimum_compilers_version(self):
...
-        if self.settings.os not in ("Linux", "Android") or tools.Version(self.version) < "0.4":
+        if self.settings.os not in ("Linux", "Android") or Version(self.version) < "0.4":
             del self.options.with_tls
 
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["CRASHPAD_ENABLE_INSTALL"] = True
+        tc.variables["CRASHPAD_ENABLE_INSTALL_DEV"] = True
+        tc.variables["CRASHPAD_ZLIB_SYSTEM"] = True
+        tc.generate()
```
Comment (uilianries) [inline recipes/sentry-crashpad/all/conanfile.py:64]:
```suggestion
        tc.generate()
        tc = CMakeDeps(self)
        tc.generate()
```

**Review by uilianries — COMMENTED** [review]
Need to generate Findxxx.cmake files.

**Comment by MartinDelille** [comment]
I guess libcurl must be updated to conan v2 before going further.

**Comment by SpaceIm** [comment]
> Any idea why pthread is not detected ?

Why are you talking about pthread? libcurl is not detected, because it's missing in requirements . It's required when SENTRY_TRANSPORT_CURL is TRUE: https://github.com/getsentry/sentry-native/blob/0.4.15/CMakeLists.txt#L268-L271

File: recipes/sentry-crashpad/all/conanfile.py [hunk 4]
```diff
@@ -53,14 +49,26 @@ def _minimum_compilers_version(self):
...
+            copy(self, patch["patch_file"], self.recipe_folder, self.export_sources_folder)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
-        if self.settings.os not in ("Linux", "Android") or tools.Version(self.version) < "0.4":
+        if self.settings.os not in ("Linux", "Android") or Version(self.version) < "0.4":
             del self.options.with_tls
 
+    def generate(self):
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:60]: move this after source()

File: recipes/sentry-crashpad/all/conanfile.py [hunk 5]
```diff
@@ -53,14 +49,26 @@ def _minimum_compilers_version(self):
     def export_sources(self):
         self.copy("CMakeLists.txt")
         for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+            copy(self, patch["patch_file"], self.recipe_folder, self.export_sources_folder)
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:52]: use export_conandata_patches

File: recipes/sentry-crashpad/all/conanfile.py [hunk 6]
```diff
@@ -53,14 +49,26 @@ def _minimum_compilers_version(self):
     def export_sources(self):
         self.copy("CMakeLists.txt")
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:50]:
```suggestion
```

File: recipes/sentry-crashpad/all/conanfile.py [hunk 7]
```diff
@@ -53,14 +49,26 @@ def _minimum_compilers_version(self):
...
+        tc.variables["CRASHPAD_ENABLE_INSTALL_DEV"] = True
+        tc.variables["CRASHPAD_ZLIB_SYSTEM"] = True
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
+
+    def layout(self):
+        cmake_layout(self)
+
     def build_requirements(self):
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:72]: to move after validate()

File: recipes/sentry-crashpad/all/conanfile.py [hunk 8]
```diff
@@ -53,14 +49,26 @@ def _minimum_compilers_version(self):
...
         if self._is_mingw:
             self.build_requires("jwasm/2.13")
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:74]:
```suggestion
            self.tool_requires("jwasm/2.13")
```

File: recipes/sentry-crashpad/all/conanfile.py [hunk 9]
```diff
@@ -74,48 +82,36 @@ def validate(self):
         if self.settings.compiler.get_safe("cppstd"):
```
Comment (SpaceIm) [inline recipes/sentry-crashpad/all/conanfile.py:82]:
```suggestion
        if self.info.settings.compiler.get_safe("cppstd"):
```

File: recipes/sentry-crashpad/all/conanfile.py (hunk 4 above)
Comment (MartinDelille) [reply recipes/sentry-crashpad/all/conanfile.py]: Done here: 0915e3156

File: recipes/sentry-crashpad/all/conanfile.py (hunk 5 above)
Comment (MartinDelille) [reply recipes/sentry-crashpad/all/conanfile.py]: Done here: 24e8c8b3b

File: recipes/sentry-crashpad/all/conanfile.py (hunk 7 above)
Comment (MartinDelille) [reply recipes/sentry-crashpad/all/conanfile.py:58]: Done here: d9e9df54a

**Comment by MartinDelille** [comment]
@prince-chrismc I think I have an issue with the package_info properties you mentioned in your workshop yesterday but I can't find out how to fix it.

**Comment by uilianries** [comment]
@MartinDelille The log says that there is only one library packaged:

```
sentry-crashpad/0.4.15 package(): Packaged 1 file: LICENSE
sentry-crashpad/0.4.15 package(): Packaged 1 '.so' file: libsentry.so
sentry-crashpad/0.4.15 package(): Packaged 1 '.h' file: sentry.h
```

But the package info expects several libraries. Thus, it's not a bug, but something is missing to generate all those libraries, or, they were installed in a wrong place.

**Comment by MartinDelille** [comment]
@uilianries I think my comprehension about CMake and packaging with conan is too basic to understand what is goind on. Could you provide me some documentation about how to deal with components with conan 2.0 ?

**Comment by uilianries** [comment]
@MartinDelille From the migration guide: https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#conanv2-properties-model and the reference: https://docs.conan.io/en/latest/migrating_to_2.0/properties.html#properties-migration

**Comment by MartinDelille** [comment]
I can't find out how to fix this issue.

**Comment by prince-chrismc** [comment]
You probably need to adjust the default values for the dirs, they change when you add a layout (2.0 migration)

I am not sure but these sound like exes? so maybe they dont need lib dirs?

https://github.com/conan-io/conan-center-index/blob/64577d26bee863e66dfb24a38723c71e00e76350/recipes/swig/all/conanfile.py#L159

or

https://github.com/conan-io/conan-center-index/blob/64577d26bee863e66dfb24a38723c71e00e76350/recipes/m4/all/conanfile.py#L110-L111

```
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_mini_chromium library 'mini_chromium' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_compat library 'crashpad_compat' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_util library 'crashpad_util' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_client library 'crashpad_client' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_snapshot library 'crashpad_snapshot' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_minidump library 'crashpad_minidump' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_handler library 'crashpad_handler_lib' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component sentry-crashpad::crashpad_tools library 'crashpad_tools' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
[HOOK - conan-center.py] post_package_info(): ERROR: [INCLUDE PATH DOES NOT EXIST (KB-H071)] Component sentry-crashpad::crashpad_mini_chromium include dir 'include/crashpad/mini_chromium' is listed in the recipe, but not found in package folder. The include dir should probably be fixed or removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H071) 
[HOOK - conan-center.py] post_package_info(): ERROR: [INCLUDE PATH DOES NOT EXIST (KB-H071)] Component sentry-crashpad::crashpad_compat include dir 'include/crashpad' is listed in the recipe, but not found in package folder. The include dir should probably be fixed or removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H071) 
```

**Comment by MartinDelille** [comment]
I checked the lib folder and there is lib involved. The issue was that I wasn't pointing the `external/crashpad/CMakeLists.txt` correctly (it was done previously by the deleted `sentry-crashpad/all/CMakeLists.txt` wrapper.

**Comment by MartinDelille** [comment]
I don't understand what is going wrong on the CI.

**Comment by SpaceIm** [comment]
> I don't understand what is going wrong on the CI.

```
ERROR: sentry-crashpad/0.4.17 (test package): Error in test() method, line 30
	handler_bin_path = os.path.join(self.dependencies["sentry-crashpad"].bin_paths[0], handler_exe)
	AttributeError: 'NoneType' object has no attribute 'dependencies'
```

`self.dependencies` is None in `test()`. I remember a discussion with @memsharded about this behavior, but I can't find the ticket where it was discussed in conan repo.

**Comment by prince-chrismc** [comment]
Oh yeah, I recall this one too... https://github.com/search?q=repo%3Aconan-io%2Fconan-center-index%20self%20dependencies%20generate&type=code 

I can not find the conversation, in my notes I wrote "the graph is only available in generate and validate" (but thats too vague to remind me) when I had worked on https://github.com/conan-io/conan-center-index/blob/34e2e2016d62a9ea3f6951a0e8e6ac65a6137e92/recipes/catch2/2.x.x/test_package/conanfile.py#L22

File: recipes/sentry-crashpad/all/conanfile.py [hunk 10]
```diff
@@ -26,96 +32,86 @@ class SentryCrashpadConan(ConanFile):
...
-        self.requires("zlib/1.2.12")
+        self.requires("libcurl/7.86.0")
+        self.requires("zlib/1.2.13")
         if self.options.get_safe("with_tls"):
-            self.requires("openssl/1.1.1n")
+            self.requires("openssl/1.1.1s")
 
     def validate(self):
-        if self.settings.compiler.get_safe("cppstd"):
+        if self.info.settings.compiler.get_safe("cppstd"):
```
Comment (prince-chrismc) [inline recipes/sentry-crashpad/all/conanfile.py:69]: This has recently changed with 1.54 and beta6 (maybe 7 its hard to keep track)

this is not longer blocked

```suggestion
        if self.settings.compiler.get_safe("cppstd"):
```

**Review by prince-chrismc — COMMENTED** [review]
This looks super good, there is one tiny nit but the conventions changes since the PR was opened, this is a green tick from me

**Comment by MartinDelille** [comment]
@prince-chrismc I can add this but I wondered how to save it ? private variable ?

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
> but I wondered how to save it ? private variable ?

I wrote docs very similar to this :) https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#passing-requirements-info-to-build

**Pull request merged**